utils/math-utils/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala

```diff
@@ -49,17 +49,102 @@ trait MathUtils {
+      override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): Number = n1.remainder(n2)
+    })(NumberTypesPromotionStrategy.ForMathOperation)
+  }
+
+  def negate(n1: Number): Number = n1 match {
+    case n1: java.lang.Byte => -n1
+    case n1: java.lang.Short => -n1
+    case n1: java.lang.Integer => -n1
+    case n1: java.lang.Long => -n1
+    case n1: java.math.BigInteger => n1.negate()
+    case n1: java.lang.Float => -n1
+    case n1: java.lang.Double => -n1
+    case n1: java.math.BigDecimal => n1.negate()
+  }
+
+  private def compare(n1: Number, n2: Number): Int = {
+    withValuesWithTheSameType(n1, n2)(new SameNumericTypeHandler[Int] {
+      override def onBytes(n1: Byte, n2: Byte): Int = n1.compareTo(n2)
+      override def onShorts(n1: Short, n2: Short): Int = n1.compareTo(n2)
+      override def onInts(n1: Int, n2: Int): Int = n1.compareTo(n2)
+      override def onLongs(n1: Long, n2: Long): Int = n1.compareTo(n2)
+      override def onBigIntegers(n1: BigInteger, n2: BigInteger): Int = n1.compareTo(n2)
+      override def onFloats(n1: Float, n2: Float): Int = n1.compareTo(n2)
+      override def onDoubles(n1: Double, n2: Double): Int = n1.compareTo(n2)
+      override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): Int = n1.compareTo(n2)
+    })(NumberTypesPromotionStrategy.ForMathOperation)
+  }
+
+  def greater(n1: Number, n2: Number): Boolean = compare(n1, n2) > 0
+  def greaterOrEqual(n1: Number, n2: Number): Boolean = compare(n1, n2) >= 0
+  def lesser(n1: Number, n2: Number): Boolean = compare(n1, n2) < 0
+  def lesserOrEqual(n1: Number, n2: Number): Boolean = compare(n1, n2) <= 0
+  def equal(n1: Number, n2: Number): Boolean = compare(n1, n2) == 0
+  def notEqual(n1: Number, n2: Number): Boolean = compare(n1, n2) != 0
+
   private def promoteThenSum(n1: Number, n2: Number)(implicit promotionStrategy: ReturningSingleClassPromotionStrategy) = {
     withNotNullValues(n1, n2) {
-      withValuesWithTheSameType(n1, n2)(new SameNumericTypeHandler {
-        override def onBytes(n1: Byte, n2: Byte): Byte = throw new IllegalStateException("Bytes should be promoted to Ints before addition")
-        override def onShorts(n1: Short, n2: Short): Short = throw new IllegalStateException("Shorts should be promoted to Ints before addition")
-        override def onInts(n1: Int, n2: Int): Int = n1 + n2
-        override def onLongs(n1: Long, n2: Long): Long = n1 + n2
-        override def onBigIntegers(n1: BigInteger, n2: BigInteger): BigInteger = n1.add(n2)
-        override def onFloats(n1: Float, n2: Float): Float = n1 + n2
-        override def onDoubles(n1: Double, n2: Double): Double = n1 + n2
-        override def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): java.math.BigDecimal = n1.add(n2)
+      withValuesWithTheSameType(n1, n2)(new SameNumericTypeHandler[Number] {
+        override def onBytes(n1: Byte, n2: Byte): Number = n1 + n2
```

Comment:
Why u removed this assertion?

Reply:
One important think is (maybe not written anywhere in comment) that those utils are in the "spel" spirit. In spel, short and bytes are promoted to int before summing which is quite reasonable for such concise languages